utils/managed_bytes.hh

```diff
@@ -30,215 +30,213 @@
-        while (index >= a->frag_size) {
-            index -= a->frag_size;
-            a = a->next;
+        const blob_storage::ref_type* a = &first_fragment();
+        while (index >= a->frag_size()) {
+            index -= a->frag_size();
+            a = &a->next();
         }
-        return a->data[index];
+        return a->data()[index];
     }
     std::unique_ptr<bytes_view::value_type[]> do_linearize_pure() const;
 
 public:
-    using size_type = blob_storage::size_type;
     struct initialized_later {};
 
     managed_bytes() {
-        _u.small.size = 0;
+        *internal_size_slot() = 0;
     }
 
-    managed_bytes(const blob_storage::char_type* ptr, size_type size)
+    managed_bytes(const blob_storage::char_type* ptr, blob_storage::size_type size)
         : managed_bytes(bytes_view(ptr, size)) {}
 
     explicit managed_bytes(const bytes& b) : managed_bytes(static_cast<bytes_view>(b)) {}
 
     template <FragmentedView View>
     explicit managed_bytes(View v);
 
-    managed_bytes(initialized_later, size_type size) {
+    managed_bytes(initialized_later, blob_storage::size_type size) {
         memory::on_alloc_point();
         if (size <= max_inline_size) {
-            _u.small.size = size;
+            *internal_size_slot() = size << 1;
         } else {
-            _u.small.size = -1;
+            if (size >= size_t(2) * 1024 * 1024 * 1024) {
+                throw std::bad_alloc();
+            }
+            write_unaligned<blob_storage::blob_storage::size_type>(external_size_slot(), net::hton((size << 1) | 1));
             auto& alctr = current_allocator();
-            auto maxseg = max_seg(alctr);
-            auto now = std::min(size_t(size), maxseg);
-            void* p = alctr.alloc(&get_standard_migrator<blob_storage>(),
-                sizeof(blob_storage) + now, alignof(blob_storage));
-            auto first = new (p) blob_storage(&_u.ptr, size, now);
-            auto last = first;
-            size -= now;
+            size_t maxalloc = max_alloc(alctr);
+            blob_storage::ref_type* last = &first_fragment();
+            new (last) blob_storage::ref_type();
             try {
-                while (size) {
-                    auto now = std::min(size_t(size), maxseg);
-                    void* p = alctr.alloc(&get_standard_migrator<blob_storage>(),
-                        sizeof(blob_storage) + now, alignof(blob_storage));
-                    last = new (p) blob_storage(&last->next, 0, now);
-                    size -= now;
+                while (size > maxalloc - sizeof(blob_storage)) {
+                    auto frag_size = maxalloc - sizeof(blob_storage) - sizeof(blob_storage::ref_type);
+                    auto p = static_cast<blob_storage*>(alctr.alloc(&get_standard_migrator<blob_storage>(), maxalloc, alignof(blob_storage)));
+                    p->backref = last;
+                    new (p->storage) blob_storage::ref_type();
+                    last->reset(p, frag_size, true);
```

Comment:
I don't understand why we `reset()` `last` with the size of the current fragment, won't that override the size of `last` with the size of this fragment?

Reply:
It's a similar issue as in your earlier comment about `backref->has_next()`. What this patch does is move the metadata about the fragment (its size) from this fragment to the previous fragment (or, in the case of first fragment, to `managed_bytes`). Hence `last` contains the metadata of this fragment, not the previous one. At this particular point in the code, `last` is zeroed out, because the fragment it describes (this one) is not created yet.